# Data Wrangling - Tools - Notebook

## Outline


* Basic Data Exploration
* Missing Values
* Duplicates
* Selecting / Dropping Columns
* String operations on whole columns
* Column splits/transformations
* MultiIndex
* Querying a Dataframe
* `DataFrame.at()` vs. `DataFrame.loc`
* Joins
* Sorting
* Grouping - Aggregation & Value Counts
* Resetting Index
* Pivot & Melt
* Column / Row Concatenation
* looping through DataFrame

In [1]:
import pandas as pd
import numpy as np

# Making copy of data

It is imporatant to make a copy of data first, as in future if we corrupt the df by accident then we must have a copy to read data.

In [2]:
data = pd.read_csv('startup_funding.csv',thousands=',')
df = data.copy()

## Basic Data Exploration

**Getting Shape of the Data**

In [3]:
df.shape

(2422, 11)

There are 2422 rows and 11 columns in the data

**Getting a list of all columns in the dataframe**

In [4]:
df.columns

Index(['index', 'SNo', 'Date', 'StartupName', 'IndustryVertical',
       'SubVertical', 'CityLocation', 'InvestorsName', 'InvestmentType',
       'AmountInUSD', 'Remarks'],
      dtype='object')

**Checking data types of all columns**

In [5]:
df.dtypes

index                 int64
SNo                   int64
Date                 object
StartupName          object
IndustryVertical     object
SubVertical          object
CityLocation         object
InvestorsName        object
InvestmentType       object
AmountInUSD         float64
Remarks              object
dtype: object

**Getting top/bottom 5 values**

By default the df.head() will show the first five rows of the data frame.

In [6]:
df.head(3)

index  SNo        Date   StartupName   IndustryVertical  \
0      0    0  01/08/2017      TouchKin         Technology   
1      1    1  02/08/2017       Ethinos         Technology   
2      2    2  02/08/2017  Leverage Edu  Consumer Internet   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   

   AmountInUSD Remarks  
0    1300000.0     NaN  
1          NaN     NaN  
2          NaN     NaN

By default the df.tail() will show the last five rows of the data frame.

In [7]:
df.tail(3)

index  SNo        Date StartupName   IndustryVertical  \
2419     47   47  28/07/2017  Gympik.com  Consumer Internet   
2420     48   48  01/06/2017     Tripeur         Technology   
2421     49   49  02/06/2017    RentOnGo          eCommerce   

                                            SubVertical CityLocation  \
2419  online marketplace for discovering fitness cen...    bangalore   
2420                   Mobile based travel ERP platform    Bangalore   
2421  Online Marketplace for Renting Bikes, Electron...    Bangalore   

                                          InvestorsName  InvestmentType  \
2419                                RoundGlass Partners    Seed Funding   
2420  Grace Grace Techno Ventures LLP, Rajul Garg & ...    Seed Funding   
2421                                  TVS Motor Company  Private Equity   

      AmountInUSD Remarks  
2419          NaN     NaN  
2420          NaN     NaN  
2421          NaN     NaN

**Getting top/bottom n values**

In [8]:
df.head(3)

index  SNo        Date   StartupName   IndustryVertical  \
0      0    0  01/08/2017      TouchKin         Technology   
1      1    1  02/08/2017       Ethinos         Technology   
2      2    2  02/08/2017  Leverage Edu  Consumer Internet   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   

   AmountInUSD Remarks  
0    1300000.0     NaN  
1          NaN     NaN  
2          NaN     NaN

In [9]:
df.tail(7)

index  SNo        Date  StartupName   IndustryVertical  \
2415     43   43  26/07/2017  ThinkerBell  Consumer Internet   
2416     44   44  27/07/2017          1mg          eCommerce   
2417     45   45  28/07/2017      Jhakaas  Consumer Internet   
2418     46   46  28/07/2017   BigStylist  Consumer Internet   
2419     47   47  28/07/2017   Gympik.com  Consumer Internet   
2420     48   48  01/06/2017      Tripeur         Technology   
2421     49   49  02/06/2017     RentOnGo          eCommerce   

                                            SubVertical CityLocation  \
2415                          Assisted Learning Startup    Bangalore   
2416                                    Online Pharmacy      Gurgaon   
2417         App-based Aggregator of Offline Businesses       Mumbai   
2418                        Beauty Services Marketplace       Mumbai   
2419  online marketplace for discovering fitness cen...    bangalore   
2420                   Mobile based travel ERP platform    Bangalore   
2421  Online Marketplace for Renting Bikes, Electron...    Bangalore   

                                          InvestorsName  InvestmentType  \
2415               Indian Angel Network, Anand Mahindra    Seed Funding   
2416  HBM Healthcare Investments, Maverick Capital V...  Private Equity   
2417                                       Amen Dhyllon    Seed Funding   
2418                              Info Edge (India) Ltd  Private Equity   
2419                                RoundGlass Partners    Seed Funding   
2420  Grace Grace Techno Ventures LLP, Rajul Garg & ...    Seed Funding   
2421                                  TVS Motor Company  Private Equity   

      AmountInUSD Remarks  
2415     200000.0     NaN  
2416   15000000.0     NaN  
2417          NaN     NaN  
2418    1250000.0     NaN  
2419          NaN     NaN  
2420          NaN     NaN  
2421          NaN     NaN

**Getting a summary of all columns**

This will display the basic stats of the **numerical columns** in the data frame

In [11]:
df.describe()

index          SNo   AmountInUSD
count  2422.000000  2422.000000  1.553000e+03
mean   1161.532205  1161.532205  1.192385e+07
std     697.598259   697.598259  6.346680e+07
min       0.000000     0.000000  1.600000e+04
25%     555.250000   555.250000  3.750000e+05
50%    1160.500000  1160.500000  1.100000e+06
75%    1765.750000  1765.750000  6.000000e+06
max    2371.000000  2371.000000  1.400000e+09

This will display the basic stats of the **object type** columns in the data frame. The 'O' means object in the following code. 

In [12]:
df.describe(include=['O'])

Date StartupName   IndustryVertical      SubVertical  \
count         2422        2422               2251             1486   
unique         701        2001                743             1364   
top     08/07/2015      Swiggy  Consumer Internet  Online Pharmacy   
freq            11           7                795               10   

       CityLocation          InvestorsName InvestmentType   Remarks  
count          2243                   2413           2421       419  
unique           71                   1885              7        69  
top       Bangalore  Undisclosed Investors   Seed Funding  Series A  
freq            649                     33           1300       177

To display all the columns at once use the following command.

In [13]:
df.describe(include='all')

index          SNo        Date StartupName   IndustryVertical  \
count   2422.000000  2422.000000        2422        2422               2251   
unique          NaN          NaN         701        2001                743   
top             NaN          NaN  08/07/2015      Swiggy  Consumer Internet   
freq            NaN          NaN          11           7                795   
mean    1161.532205  1161.532205         NaN         NaN                NaN   
std      697.598259   697.598259         NaN         NaN                NaN   
min        0.000000     0.000000         NaN         NaN                NaN   
25%      555.250000   555.250000         NaN         NaN                NaN   
50%     1160.500000  1160.500000         NaN         NaN                NaN   
75%     1765.750000  1765.750000         NaN         NaN                NaN   
max     2371.000000  2371.000000         NaN         NaN                NaN   

            SubVertical CityLocation          InvestorsName InvestmentType  \
count              1486         2243                   2413           2421   
unique             1364           71                   1885              7   
top     Online Pharmacy    Bangalore  Undisclosed Investors   Seed Funding   
freq                 10          649                     33           1300   
mean                NaN          NaN                    NaN            NaN   
std                 NaN          NaN                    NaN            NaN   
min                 NaN          NaN                    NaN            NaN   
25%                 NaN          NaN                    NaN            NaN   
50%                 NaN          NaN                    NaN            NaN   
75%                 NaN          NaN                    NaN            NaN   
max                 NaN          NaN                    NaN            NaN   

         AmountInUSD   Remarks  
count   1.553000e+03       419  
unique           NaN        69  
top              NaN  Series A  
freq             NaN       177  
mean    1.192385e+07       NaN  
std     6.346680e+07       NaN  
min     1.600000e+04       NaN  
25%     3.750000e+05       NaN  
50%     1.100000e+06       NaN  
75%     6.000000e+06       NaN  
max     1.400000e+09       NaN

**Getting unique values of  single column**

After running this command you will see that it returns all unique values that are present in the column **InvestmentType**.

Note that **PrivateEquity** and **Private Equity** are two unique values.

In [15]:
df['InvestmentType'].unique()

array(['Private Equity', 'Seed Funding', 'Debt Funding', nan,
       'SeedFunding', 'PrivateEquity', 'Crowd funding', 'Crowd Funding'],
      dtype=object)

## Missing Values

**Checking which columns have missing values and how many**

In [14]:
df.isnull().sum()

index                  0
SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation         179
InvestorsName          9
InvestmentType         1
AmountInUSD          869
Remarks             2003
dtype: int64

**Treating Missing Values of a single column (Series)**

In [15]:
df['AmountInUSD'] = df['AmountInUSD'].fillna(0)

After running the following command you'll see that AmountInUSD is now filled with 0 and it doesn't have any missing/null values.

In [16]:
df.isnull().sum()

index                  0
SNo                    0
Date                   0
StartupName            0
IndustryVertical     171
SubVertical          936
CityLocation         179
InvestorsName          9
InvestmentType         1
AmountInUSD            0
Remarks             2003
dtype: int64

**Treating all Missing Values at once**

In [19]:
df.fillna('0', inplace=True)

The **inplace=True** parameter in the above code acts like an assignment operator and re-assigns the the changed changed df.

In [20]:
df.isnull().sum()

index               0
SNo                 0
Date                0
StartupName         0
IndustryVertical    0
SubVertical         0
CityLocation        0
InvestorsName       0
InvestmentType      0
AmountInUSD         0
Remarks             0
dtype: int64

Now you can see that all the columns have no **0** or **null values**

## Duplicates

**Checking for duplicates in a specefic column**

In [21]:
df.duplicated(['StartupName']).sum()

421

This means that there are **421** StartupName instances that are exactly the same.





**Checking for whole row duplicates**

In [22]:
df.duplicated().sum()

50

this returns the total number of rows that are duplicated.

In [23]:
df[df.duplicated()]

index  SNo        Date            StartupName   IndustryVertical  \
2372      0    0  01/08/2017               TouchKin         Technology   
2373      1    1  02/08/2017                Ethinos         Technology   
2374      2    2  02/08/2017           Leverage Edu  Consumer Internet   
2375      3    3  02/08/2017                   Zepo  Consumer Internet   
2376      4    4  02/08/2017           Click2Clinic  Consumer Internet   
2377      5    5  01/07/2017          Billion Loans  Consumer Internet   
2378      6    6  03/07/2017       Ecolibriumenergy         Technology   
2379      7    7  04/07/2017                  Droom          eCommerce   
2380      8    8  05/07/2017              Jumbotail          eCommerce   
2381      9    9  05/07/2017                 Moglix          eCommerce   
2382     10   10  05/07/2017             Timesaverz  Consumer Internet   
2383     11   11  06/07/2017                 Minjar         Technology   
2384     12   12  06/07/2017            MyCity4kids  Consumer Internet   
2385     13   13  07/07/2017               Clip App  Consumer Internet   
2386     14   14  07/07/2017            Upwardly.in  Consumer Internet   
2387     15   15  10/07/2017             Autorox.co         Technology   
2388     16   16  11/07/2017                 Fabogo  Consumer Internet   
2389     17   17  11/07/2017             Flickstree  Consumer Internet   
2390     18   18  11/07/2017            Design Cafe  Consumer Internet   
2391     19   19  12/07/2017               Innoviti         Technology   
2392     20   20  12/07/2017               VDeliver          Logistics   
2393     21   21  12/07/2017               Bottr.me         Technology   
2394     22   22  12/07/2017               Arcatron         Technology   
2395     23   23  14/07/2017               QwikSpec         Technology   
2396     24   24  14/07/2017                Chumbak          eCommerce   
2397     25   25  17/07/2017                Increff         Technology   
2398     26   26  17/07/2017                 Vayana         Technology   
2399     27   27  18/07/2017              MObiquest         Technology   
2400     28   28  18/07/2017                  Ambee  Consumer Internet   
2401     29   29  18/07/2017        Ideal Insurance  Consumer Internet   
2402     30   30  18/07/2017  Hypernova Interactive         Technology   
2403     31   31  19/07/2017              Rentomojo  Consumer Internet   
2404     32   32  19/07/2017                 AirCTO  Consumer Internet   
2405     33   33  19/07/2017               Playablo         Technology   
2406     34   34  20/07/2017                 Trupay         Technology   
2407     35   35  21/07/2017             Brick2Wall          eCommerce   
2408     36   36  21/07/2017            FableStreet          eCommerce   
2409     37   37  21/07/2017        Monsoon Fintech         Technology   
2410     38   38  21/07/2017              MonkeyBox  Consumer Internet   
2411     39   39  25/07/2017            Noticeboard  Consumer Internet   
2412     40   40  25/07/2017                 Byju’s  Consumer Internet   
2413     41   41  26/07/2017      Creator’s Gurukul             Others   
2414     42   42  26/07/2017             Fab Hotels  Consumer Internet   
2415     43   43  26/07/2017            ThinkerBell  Consumer Internet   
2416     44   44  27/07/2017                    1mg          eCommerce   
2417     45   45  28/07/2017                Jhakaas  Consumer Internet   
2418     46   46  28/07/2017             BigStylist  Consumer Internet   
2419     47   47  28/07/2017             Gympik.com  Consumer Internet   
2420     48   48  01/06/2017                Tripeur         Technology   
2421     49   49  02/06/2017               RentOnGo          eCommerce   

                                            SubVertical CityLocation  \
2372                           Predictive Care Platform    Bangalore   
2373                           Digital Marketing Agency       Mumbai   
2374      Onlin

This shows that these are the rows that are duplicated. They may have more than 2 instances.

**Deleting duplicates from a specefic column**

In [24]:
df.shape

(2422, 11)

Following code will delete the duplicates that are present in **StartupName**. The **keep='First'** parameter keeps the first occurance of the instance that is duplicated and deletes the rest. So in resutl only unique instances are left

In [25]:
df.drop_duplicates(['StartupName'], keep='first').shape # Default is keep=first

(2001, 11)

In [26]:
df.drop_duplicates(['StartupName'], keep='last').shape #keep=last keeps the last instance and deletes the rest

(2001, 11)

In [27]:
df.drop_duplicates(['StartupName'], keep=False).shape # this deletes all the occurances of the duplicates so none is left.

(1679, 11)

**NOTE:** In all the above occurances neither did we use the inplace=True parameter nor manually assigned it. So there won't be any change in the df

**Deleting whole row duplicates**

In [28]:
df.drop_duplicates().shape

(2372, 11)

In [29]:
df.shape

(2422, 11)

Now we will use the inplace=True parameter to save the data frame

In [30]:
df.drop_duplicates(inplace=True)
df.shape

(2372, 11)

## Selecting / Dropping Columns

**Selecting Columns**

In order to select multiple columns we have to pass their names as list in the data frame

In [31]:
df[['Date', 'CityLocation', 'AmountInUSD']]

Date CityLocation  AmountInUSD
0      01/08/2017    Bangalore    1300000.0
1      02/08/2017       Mumbai          0.0
2      02/08/2017    New Delhi          0.0
3      02/08/2017       Mumbai     500000.0
4      02/08/2017    Hyderabad     850000.0
5      01/07/2017    Bangalore    1000000.0
6      03/07/2017    Ahmedabad    2600000.0
7      04/07/2017      Gurgaon   20000000.0
8      05/07/2017    Bangalore    8500000.0
9      05/07/2017        Noida   12000000.0
10     05/07/2017       Mumbai    1000000.0
11     06/07/2017    Bangalore          0.0
12     06/07/2017      Gurgaon          0.0
13     07/07/2017    Bangalore    1000000.0
14     07/07/2017    Bangalore          0.0
15     10/07/2017    Hyderabad    3000000.0
16     11/07/2017         Pune    2250000.0
17     11/07/2017       Mumbai     464000.0
18     11/07/2017    Bangalore          0.0
19     12/07/2017    Bangalore   18500000.0
20     12/07/2017    Hyderabad          0.0
21     12/07/2017    Bangalore          0.0
22     12/07/2017         Pune          0.0
23     14/07/2017    Bangalore     540000.0
24     14/07/2017    Bangalore    1700000.0
25     17/07/2017    Bangalore    2000000.0
26     17/07/2017         Pune    4000000.0
27     18/07/2017        Noida          0.0
28     18/07/2017    Hyderabad          0.0
29     18/07/2017      Kolkata          0.0
...           ...          ...          ...
2342   19/01/2015            0     330000.0
2343   19/01/2015            0          0.0
2344   19/01/2015            0     200000.0
2345   19/01/2015            0      16000.0
2346   19/01/2015            0      16000.0
2347   19/01/2015            0      16000.0
2348   19/01/2015            0      16000.0
2349   19/01/2015            0      16000.0
2350   20/01/2015            0   11000000.0
2351   20/01/2015            0          0.0
2352   20/01/2015            0     200000.0
2353   21/01/2015            0   14750000.0
2354   21/01/2015            0     325000.0
2355   21/05/2015            0          0.0
2356   21/05/2015            0    3250000.0
2357  22/01//2015            0     200000.0
2358   22/01/2015            0          0.0
2359   22/01/2015            0          0.0
2360   22/01/2015            0   56000000.0
2361   22/01/2015            0   31000000.0
2362   24/01/2015            0    1500000.0
2363   24/01/2015            0          0.0
2364   25/01/2015            0    1000000.0
2365   27/01/2015            0   50000000.0
2366   28/01/2015            0    2500000.0
2367   29/01/2015            0    4500000.0
2368   29/01/2015            0     825000.0
2369   30/01/2015            0    1500000.0
2370   30/01/2015            0          0.0
2371   31/01/2015            0     140000.0

[2372 rows x 3 columns]

**Dropping Columns**

In [32]:
df.head()

index  SNo        Date   StartupName   IndustryVertical  \
0      0    0  01/08/2017      TouchKin         Technology   
1      1    1  02/08/2017       Ethinos         Technology   
2      2    2  02/08/2017  Leverage Edu  Consumer Internet   
3      3    3  02/08/2017          Zepo  Consumer Internet   
4      4    4  02/08/2017  Click2Clinic  Consumer Internet   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   
3                         DIY Ecommerce platform       Mumbai   
4                  healthcare service aggregator    Hyderabad   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   
3  Kunal Shah, LetsVenture, Anupam Mittal, Hetal ...    Seed Funding   
4                     Narottam Thudi, Shireesh Palle    Seed Funding   

   AmountInUSD Remarks  
0    1300000.0       0  
1          0.0       0  
2          0.0       0  
3     500000.0       0  
4     850000.0       0

In [33]:
df.drop('index', axis=1).shape

(2372, 10)

In [34]:
df.shape

(2372, 11)

use the inplace true paramter to save the data frame

In [35]:
df.drop('index', axis=1,inplace=True)

In [36]:
df.shape

(2372, 10)

In [37]:
df.head()

SNo        Date   StartupName   IndustryVertical  \
0    0  01/08/2017      TouchKin         Technology   
1    1  02/08/2017       Ethinos         Technology   
2    2  02/08/2017  Leverage Edu  Consumer Internet   
3    3  02/08/2017          Zepo  Consumer Internet   
4    4  02/08/2017  Click2Clinic  Consumer Internet   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   
3                         DIY Ecommerce platform       Mumbai   
4                  healthcare service aggregator    Hyderabad   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   
3  Kunal Shah, LetsVenture, Anupam Mittal, Hetal ...    Seed Funding   
4                     Narottam Thudi, Shireesh Palle    Seed Funding   

   AmountInUSD Remarks  
0    1300000.0       0  
1          0.0       0  
2          0.0       0  
3     500000.0       0  
4     850000.0       0

## String Operations on whole columns

**String Replacement**

In [38]:
df['InvestmentType'].unique()

array(['Private Equity', 'Seed Funding', 'Debt Funding', '0',
       'SeedFunding', 'PrivateEquity', 'Crowd funding', 'Crowd Funding'],
      dtype=object)

We want to replace *PrivateEquatiy* which is written together to be written as *Private Equatiy*

In [39]:
# first parameter in str.replace is the one which we want to replace and the second one is with which we are replacing it
df['InvestmentType'].str.replace('PrivateEquity', 'Private Equity').unique() 

array(['Private Equity', 'Seed Funding', 'Debt Funding', '0',
       'SeedFunding', 'Crowd funding', 'Crowd Funding'], dtype=object)

In [40]:
df['InvestmentType'].unique()

array(['Private Equity', 'Seed Funding', 'Debt Funding', '0',
       'SeedFunding', 'PrivateEquity', 'Crowd funding', 'Crowd Funding'],
      dtype=object)

As you can see it is back in the df because we did not re assign it. **Inplace=True** does **NOT** work here so we have to reassign it manually

In [41]:
df['InvestmentType'] = df['InvestmentType'].str.replace('PrivateEquity', 'Private Equity')
df['InvestmentType'] = df['InvestmentType'].str.replace('SeedFunding', 'Seed Funding')
df['InvestmentType'] = df['InvestmentType'].str.replace('Crowd funding', 'Crowd Funding')
df['InvestmentType'] = df['InvestmentType'].str.replace('0', 'Other')

In [42]:
df['InvestmentType'].unique()

array(['Private Equity', 'Seed Funding', 'Debt Funding', 'Other',
       'Crowd Funding'], dtype=object)

**Capitalization**

To capitalize every name in a specific column use the follwing code

In [43]:
df['CityLocation'].str.lower()

0       bangalore
1          mumbai
2       new delhi
3          mumbai
4       hyderabad
5       bangalore
6       ahmedabad
7         gurgaon
8       bangalore
9           noida
10         mumbai
11      bangalore
12        gurgaon
13      bangalore
14      bangalore
15      hyderabad
16           pune
17         mumbai
18      bangalore
19      bangalore
20      hyderabad
21      bangalore
22           pune
23      bangalore
24      bangalore
25      bangalore
26           pune
27          noida
28      hyderabad
29        kolkata
          ...    
2342            0
2343            0
2344            0
2345            0
2346            0
2347            0
2348            0
2349            0
2350            0
2351            0
2352            0
2353            0
2354            0
2355            0
2356            0
2357            0
2358            0
2359            0
2360            0
2361            0
2362            0
2363            0
2364            0
2365            0
2366      

**Checking if there is a substring in each column value**

In [44]:
df[df['InvestorsName'].str.contains('Khan')]

SNo        Date         StartupName                 IndustryVertical  \
81      81  16/06/2017             Fincash                Consumer Internet   
780    780  17/08/2016           MaalGaadi                        Logistics   
871    871  15/07/2016             BaggOut                        eCommerce   
920    920  10/06/2016      Kickstart Jobs                       Technology   
947    947  17/06/2016                 BYG                Consumer Internet   
1094  1094  13/04/2016         Legalraasta                Consumer Internet   
1155  1155   03/3/2016  Imarticus Learning                        Education   
1581  1581  19/11/2015       PlaceofOrigin  Online Gourmet Food Marketplace   
1598  1598  25/11/2015              Tooler   On Demand Laundry Services App   
1793  1793  29/09/2015          LoanCircle     Consumer lending marketplace   
1905  1905  02/07/2015             Zenatix      Internet of Things platform   

                                            SubVertical CityLocation  \
81                            Personal Finance platform       Mumbai   
780                        Online Logistics Marketplace       Indore   
871                             Women’s Fashion etailer    New Delhi   
920                         Entry level hiring platform      Gurgaon   
947       Fitness centre Discovery & Booking Mobile app    Bangalore   
1094                 Online legal Services for Startups    New Delhi   
1155  Financial Services & Analytics Education Insti...       Mumbai   
1581                                                  0    Bangalore   
1598                                                  0    New Delhi   
1793                                                  0    Bangalore   
1905                                                  0      Gurgaon   

                                          InvestorsName  InvestmentType  \
81               Mohammed Khan, Sameer Narayan & Others    Seed Funding   
780     Swan Angel Network,Sachin Khandelwal and others    Seed Funding   
871   Sumit Jain, Sumit Jain, Anurag Gupta, Varun Kh...    Seed Funding   
920   Vivek Joshi, Mohit Satyanand, Amit Banati, Aru...    Seed Funding   
947             Sanjay Verma, Amit Khanna (LetsVenture)    Seed Funding   
1094                Pravin Khandelwal, Yatin Kumar Jain    Seed Funding   
1155  Blinc Advisors, Amit Nanavati, Tashwinder Sing...  Private Equity   
1581  S.D. Shibulal, Kris Gopalakrishnan, Srinath Ba...    Seed Funding   
1598                         Raghu Khanna, Sameer Gupta    Seed Funding   
1793              Zishaan Hayath, Rahul Khanna & Others    Seed Funding   
1905  Rajan Anandan, Kunal Bahl, Rohit Bansal, Rahul...    Seed Funding   

      AmountInUSD Remarks  
81       100000.0       0  
780      375000.0       0  
871           0.0       0  
920           0.0       0  
947           0.0       0  
1094    1000000.0       0  
1155    1000000.0       0  
1581          0.0       0  
1598     110000.0       0  
1793          0.0       0  
1905     161000.0       0

In [45]:
df[df['InvestorsName'].str.contains('Khan')]

SNo        Date         StartupName                 IndustryVertical  \
81      81  16/06/2017             Fincash                Consumer Internet   
780    780  17/08/2016           MaalGaadi                        Logistics   
871    871  15/07/2016             BaggOut                        eCommerce   
920    920  10/06/2016      Kickstart Jobs                       Technology   
947    947  17/06/2016                 BYG                Consumer Internet   
1094  1094  13/04/2016         Legalraasta                Consumer Internet   
1155  1155   03/3/2016  Imarticus Learning                        Education   
1581  1581  19/11/2015       PlaceofOrigin  Online Gourmet Food Marketplace   
1598  1598  25/11/2015              Tooler   On Demand Laundry Services App   
1793  1793  29/09/2015          LoanCircle     Consumer lending marketplace   
1905  1905  02/07/2015             Zenatix      Internet of Things platform   

                                            SubVertical CityLocation  \
81                            Personal Finance platform       Mumbai   
780                        Online Logistics Marketplace       Indore   
871                             Women’s Fashion etailer    New Delhi   
920                         Entry level hiring platform      Gurgaon   
947       Fitness centre Discovery & Booking Mobile app    Bangalore   
1094                 Online legal Services for Startups    New Delhi   
1155  Financial Services & Analytics Education Insti...       Mumbai   
1581                                                  0    Bangalore   
1598                                                  0    New Delhi   
1793                                                  0    Bangalore   
1905                                                  0      Gurgaon   

                                          InvestorsName  InvestmentType  \
81               Mohammed Khan, Sameer Narayan & Others    Seed Funding   
780     Swan Angel Network,Sachin Khandelwal and others    Seed Funding   
871   Sumit Jain, Sumit Jain, Anurag Gupta, Varun Kh...    Seed Funding   
920   Vivek Joshi, Mohit Satyanand, Amit Banati, Aru...    Seed Funding   
947             Sanjay Verma, Amit Khanna (LetsVenture)    Seed Funding   
1094                Pravin Khandelwal, Yatin Kumar Jain    Seed Funding   
1155  Blinc Advisors, Amit Nanavati, Tashwinder Sing...  Private Equity   
1581  S.D. Shibulal, Kris Gopalakrishnan, Srinath Ba...    Seed Funding   
1598                         Raghu Khanna, Sameer Gupta    Seed Funding   
1793              Zishaan Hayath, Rahul Khanna & Others    Seed Funding   
1905  Rajan Anandan, Kunal Bahl, Rohit Bansal, Rahul...    Seed Funding   

      AmountInUSD Remarks  
81       100000.0       0  
780      375000.0       0  
871           0.0       0  
920           0.0       0  
947           0.0       0  
1094    1000000.0       0  
1155    1000000.0       0  
1581          0.0       0  
1598     110000.0       0  
1793          0.0       0  
1905     161000.0       0

In [46]:
df['InvestorsName'].str.contains('Khan').sum()

11

In [47]:
df['InvestorsName'].str.contains('khan').sum()

3

**What do you observe?**
<br>The str.contains() function is case sensative. If you pass Khan with capital K then it will search it with capital K otherwise if you pass lower k then it will search accordingly.

## Column Transformations / Splits

**Changing Data types**

___1. Normal conversion___

In [48]:
df.dtypes

SNo                   int64
Date                 object
StartupName          object
IndustryVertical     object
SubVertical          object
CityLocation         object
InvestorsName        object
InvestmentType       object
AmountInUSD         float64
Remarks              object
dtype: object

In [49]:
df['AmountInUSD']=df['AmountInUSD'].astype(np.int32)

In [50]:
df.dtypes

SNo                  int64
Date                object
StartupName         object
IndustryVertical    object
SubVertical         object
CityLocation        object
InvestorsName       object
InvestmentType      object
AmountInUSD          int32
Remarks             object
dtype: object

___2. Dates Conversion___

This will return an error as there are unknown string formats present. The below function will only work when date is in a specific format like 1/1/2001

In [51]:
pd.to_datetime(df['Date'])

ValueError: ('Unknown string format:', '12/05.2015')

we saw in the data that there were some anomolies present. Like in the dates double slashes were used instead of single so we had to replace them
<br> The errors='coerce' parameter ignores the errors and converts the rest.

In [ ]:
df['Date'] = pd.to_datetime(df['Date'].str.replace('//', '/'), dayfirst=True, errors='coerce')

In [ ]:
df.dtypes

**Getting months/year/days as separate columns from a datetime column**

In [ ]:
df['month'] = pd.DatetimeIndex(df['Date']).month

**Exercise:** Extract day and year

**Extracting new columns from existing string columns**

In [ ]:
df['InvestmentType'].str.split(' ')

In [ ]:
df['InvestmentType'].str.split(' ').str[0]

## MultiIndexing

In [ ]:
grouped = df.groupby(['IndustryVertical', 'SubVertical'])

In [ ]:
new = grouped.agg({'AmountInUSD': {'Mean': np.mean, 'Sum': np.sum}})

In [ ]:
new

In [ ]:
new['AmountInUSD']

#### You can slice a MultiIndex by providing multiple indexers.


You can use pandas.IndexSlice to facilitate a more natural syntax using :,



In [ ]:
idx = pd.IndexSlice
new.loc[idx[:, ['AI Based Personal Assistant', 'App based cab aggregator']], idx[:, 'Sum']]


In [ ]:
new['AmountInUSD']['Sum']

## Querying a Dataframe

**Querying based on a single value**

In [ ]:
df['AmountInUSD'] < 100000

In [ ]:
df[df['AmountInUSD'] < 100000]

**Querying based on multiple columns**

In [ ]:
df[(df['AmountInUSD'] < 100000) & (df['CityLocation'] == 'New Delhi')]

**Querying based on a list of values**

In [ ]:
df['CityLocation'].isin(['New Delhi', 'Bangalore', 'Mumbai'])

In [ ]:
#df[(df['CityLocation'].isin(['New Delhi', 'Bangalore', 'Mumbai'])) & (df['AmountInUSD'] < 100000)]

df[
    (df['AmountInUSD'] > 5000) 
    &
    (df['CityLocation'].isin(['New Dehli']))
]



## `DataFrame.at` vs. `DataFrame.loc`

In [74]:
df.at[133,['StartupName', 'AmountInUSD']]

TypeError: unhashable type: 'list'

In [75]:
df.at[133,'StartupName']

'Curie Labs'

In [76]:
df.loc[133,['StartupName', 'AmountInUSD']]

StartupName    Curie Labs
AmountInUSD         50000
Name: 133, dtype: object

In [77]:
%timeit df.at[133,'StartupName']

9 µs ± 2.09 µs per loop (mean ± std. dev. of 7 runs, 100000 loops each)


In [78]:
%timeit df.loc[133,'StartupName']

12.7 µs ± 2.05 µs per loop (mean ± std. dev. of 7 runs, 10000 loops each)


## Joins

In [79]:
rows1 = df[0:10]
rows2 = df[5:12]
print(rows1.shape)
rows1
table1 = rows1[['SNo', 'Date', 'CityLocation']]
table2 = rows2[['SNo', 'StartupName', 'IndustryVertical', 'InvestmentType']]
print(table1.shape)
table1

(10, 11)
(10, 3)


SNo       Date CityLocation
0    0 2017-08-01    Bangalore
1    1 2017-08-02       Mumbai
2    2 2017-08-02    New Delhi
3    3 2017-08-02       Mumbai
4    4 2017-08-02    Hyderabad
5    5 2017-07-01    Bangalore
6    6 2017-07-03    Ahmedabad
7    7 2017-07-04      Gurgaon
8    8 2017-07-05    Bangalore
9    9 2017-07-05        Noida

In [80]:
print(table2.shape)
table2

(7, 4)


SNo       StartupName   IndustryVertical  InvestmentType
5     5     Billion Loans  Consumer Internet    Seed Funding
6     6  Ecolibriumenergy         Technology  Private Equity
7     7             Droom          eCommerce  Private Equity
8     8         Jumbotail          eCommerce  Private Equity
9     9            Moglix          eCommerce  Private Equity
10   10        Timesaverz  Consumer Internet  Private Equity
11   11            Minjar         Technology    Seed Funding

In [81]:
table1.merge(table2, how='inner', on='SNo')

SNo       Date CityLocation       StartupName   IndustryVertical  \
0    5 2017-07-01    Bangalore     Billion Loans  Consumer Internet   
1    6 2017-07-03    Ahmedabad  Ecolibriumenergy         Technology   
2    7 2017-07-04      Gurgaon             Droom          eCommerce   
3    8 2017-07-05    Bangalore         Jumbotail          eCommerce   
4    9 2017-07-05        Noida            Moglix          eCommerce   

   InvestmentType  
0    Seed Funding  
1  Private Equity  
2  Private Equity  
3  Private Equity  
4  Private Equity

## Sorting

In [82]:
df.sort_values(by=['CityLocation', 'IndustryVertical'])

SNo       Date                        StartupName  \
2201  2201 2015-03-02                     TrulyMadly.com   
2202  2202 2015-03-02                        InstaLively   
2203  2203 2015-03-03                              Vserv   
2204  2204 2015-03-03                         Intruo.com   
2205  2205 2015-03-05                            Niffler   
2206  2206 2015-03-05                  CustomFurnish.com   
2207  2207 2015-03-06                        MapMyGenome   
2208  2208 2015-03-09  Crowdfire (formerly Justunfollow)   
2209  2209 2015-03-09                          Bite Club   
2210  2210 2015-03-10                         Localbanya   
2211  2211 2015-03-11                             Appedo   
2212  2212 2015-03-11                         Fitraq.com   
2213  2213 2015-03-11                          ChargeBee   
2214  2214 2015-03-11                      Groupon India   
2215  2215 2015-03-12                          Process 9   
2216  2216 2015-03-12                      MoonFrog Labs   
2217  2217 2015-03-13                          Foodpanda   
2218  2218 2015-03-13                              Paytm   
2219  2219 2015-03-13                        DiabetOmics   
2220  2220 2015-03-13                                Qyk   
2221  2221 2015-03-14                       Daily Rounds   
2222  2222 2015-03-14                            Orobind   
2223  2223 2015-03-14                            Routofy   
2224  2224 2015-03-16                     Digital Gorkha   
2225  2225 2015-03-16                           NestAway   
2226  2226 2015-03-16                   The Better India   
2227  2227 2015-03-16                          Meru Cabs   
2228  2228 2015-03-17                            GolfLAN   
2229  2229 2015-03-18                        Drivestream   
2230  2230 2015-03-18                             wWhere   
...    ...        ...                                ...   
1068  1068 2016-04-01                            Dil Mil   
1152  1152 2016-03-02                            Skedool   
1417  1417 2016-01-27                         Teabox.com   
915    915 2016-06-09                             Teabox   
454    454 2016-12-07                           LookAtMe   
1489  1489 2015-12-15                   Deal Street Asia   
1620  1620 2015-10-01                             Muvizz   
538    538 2016-11-14                          Active.ai   
1005  1005 2016-05-04                          Active.ai   
290    290 2017-03-16                           ShilpMIS   
1039  1039 2016-05-18                       PurpleHealth   
125    125 2017-05-05                Good Methods Global   
836    836 2016-07-04                          Fulfil.IO   
1742  1742 2015-09-14                   UE Life Sciences   
940    940 2016-06-16                              Saavn   
1534  1534 2015-11-03                            Bidgely   
2005  2005 2015-06-02                   Karma Healthcare   
2168  2168 2015-04-17                           Robosoft   
432    432 2017-01-02                            99Games   
583    583 2016-11-30                         Math Buddy   
713    713 2016-09-21                         PurpleDocs   
935    935 2016-06-15                         Oneway.cab   
996    996 2016-05-03                           BoiBanit   
1249  1249 2016-02-05                        Gingercrush   
2171  2171 2015-04-20                          Pastiwala   
1530  1530 2015-11-03                           boibanit   
1555  1555 2015-11-11                        Gingercrush   
259    259 2017-03-02                     Kreate Konnect   
1251  1251 2016-02-05                          DawaiLelo   
47      47 2017-07-28                         Gympik.com   

                            IndustryVertical  \
2201                                       0   
2202                                       0   
2203                                       0   
2204                                       0   
2205                                       0  

In [83]:
df.sort_values(by=['CityLocation', 'IndustryVertical'], ascending=False)

SNo       Date          StartupName  \
47      47 2017-07-28           Gympik.com   
1251  1251 2016-02-05            DawaiLelo   
259    259 2017-03-02       Kreate Konnect   
1555  1555 2015-11-11          Gingercrush   
1530  1530 2015-11-03             boibanit   
2171  2171 2015-04-20            Pastiwala   
1249  1249 2016-02-05          Gingercrush   
583    583 2016-11-30           Math Buddy   
713    713 2016-09-21           PurpleDocs   
935    935 2016-06-15           Oneway.cab   
996    996 2016-05-03             BoiBanit   
432    432 2017-01-02              99Games   
2168  2168 2015-04-17             Robosoft   
2005  2005 2015-06-02     Karma Healthcare   
1534  1534 2015-11-03              Bidgely   
940    940 2016-06-16                Saavn   
1742  1742 2015-09-14     UE Life Sciences   
836    836 2016-07-04            Fulfil.IO   
125    125 2017-05-05  Good Methods Global   
1039  1039 2016-05-18         PurpleHealth   
290    290 2017-03-16             ShilpMIS   
538    538 2016-11-14            Active.ai   
1005  1005 2016-05-04            Active.ai   
1620  1620 2015-10-01               Muvizz   
1489  1489 2015-12-15     Deal Street Asia   
454    454 2016-12-07             LookAtMe   
915    915 2016-06-09               Teabox   
1417  1417 2016-01-27           Teabox.com   
1152  1152 2016-03-02              Skedool   
1068  1068 2016-04-01              Dil Mil   
...    ...        ...                  ...   
2342  2342 2015-01-19              Refiral   
2343  2343 2015-01-19         Ping Digital   
2344  2344 2015-01-19      HealthyWorld.in   
2345  2345 2015-01-19         Hostel Dunia   
2346  2346 2015-01-19      Play your sport   
2347  2347 2015-01-19              Yo Grad   
2348  2348 2015-01-19               Enabli   
2349  2349 2015-01-19                  CBS   
2350  2350 2015-01-20            Hopscotch   
2351  2351 2015-01-20     Amagi Media Labs   
2352  2352 2015-01-20      HealthyWorld.in   
2353  2353 2015-01-21          Simplilearn   
2354  2354 2015-01-21         MyDreamStore   
2355  2355 2015-05-21                 Knit   
2356  2356 2015-05-21              Villgro   
2357  2357 2015-01-22         Corporate360   
2358  2358 2015-01-22            Freshmonk   
2359  2359 2015-01-22      Englishleap.com   
2360  2360 2015-01-22               Antuit   
2361  2361 2015-01-22        Caratlane.com   
2362  2362 2015-01-24           Aasaanjobs   
2363  2363 2015-01-24             Impartus   
2364  2364 2015-01-25           Thrill App   
2365  2365 2015-01-27         CarDekho.com   
2366  2366 2015-01-28        Grabhouse.com   
2367  2367 2015-01-29           Printvenue   
2368  2368 2015-01-29             Graphene   
2369  2369 2015-01-30       Mad Street Den   
2370  2370 2015-01-30            Simplotel   
2371  2371 2015-01-31     couponmachine.in   

                            IndustryVertical  \
47                         Consumer Internet   
1251                       Consumer Internet   
259                               Technology   
1555          Product Customization Platform   
1530        Online Food Ordering Marketplace   
2171         Material Collection & Recycling   
1249                               Ecommerce   
583                        Consumer Internet   
713                        Consumer Internet   
935                        Consumer Internet   
996                        Consumer Internet   
432                               Technology   
2168                  Mobile App Development   
2005  Tech-enabled Rural Healthcare Services   
1534    Efficient Energy Management platform   
940                        Consumer Internet   
1742        Early Cancer Detection Solutions   
836                        Consumer Internet   
125                               Technology   
1039                       Consumer Internet   
290                               Technology   
538                               Technology   
1005                              Technology   
16

## Grouping - Aggregation & Value Counts

**Without grouping**

In [84]:
df['AmountInUSD'].sum()

18347386476

In [85]:
df['IndustryVertical'].unique()

array(['Technology', 'Consumer Internet', 'eCommerce', 'Logistics',
       'Others', 'Healthcare', 'Food & Beverage', 'ECommerce', 'Finance',
       'Education', 'Food & Beverages', 'ecommerce', 'healthcare',
       'Real Estate', 'FMCG', 'Reality', 'Auto', 'Ecommerce', 'BFSI',
       'Consumer Interne', 'Online Education Information platform',
       'Brand Licensing Startup',
       'Gourmet Food Discovery & Delivery platform',
       'Transportation & Logistics Platform',
       'Enterprise Marketing Automation platform',
       'Health, Wellness & Beauty Services App', 'Digital Healthcare',
       'Last Minute Hotel Booking App', 'Womens Fashion Wear Portal',
       'Product Learning platform',
       'Online Food ordering & Delivery platform',
       'App based Bus Pooling Services', 'Social Learning Platform',
       'Social Fitness platform', 'On Demand Mobile app developer',
       'Car Maintenance & Management mobile app',
       'Online Wedding Marketplace', 'Splitting Bills 

In [86]:
df['IndustryVertical'].value_counts()

Consumer Internet                                          772
Technology                                                 313
eCommerce                                                  171
0                                                          171
ECommerce                                                   53
Healthcare                                                  30
Logistics                                                   24
Education                                                   20
Food & Beverage                                             19
Finance                                                      9
Others                                                       6
Online Food Delivery                                         5
Online Education Platform                                    5
Real Estate                                                  4
FMCG                                                         3
ecommerce                                              

**Grouping based on a single column**

In [87]:
df.groupby('InvestmentType')


In [88]:
df.groupby('InvestmentType')['AmountInUSD'].sum()

InvestmentType
Crowd Funding     1.557680e+05
Debt Funding      7.800000e+06
Other             0.000000e+00
Private Equity    1.800708e+10
Seed Funding      3.323457e+08
Name: AmountInUSD, dtype: float64

In [89]:
df.groupby('CityLocation')['IndustryVertical'].value_counts()

CityLocation         IndustryVertical                                  
0                    0                                                     171
                     Consumer Internet                                       5
                     Mobile Point of Sale payment solution                   1
                     Online Travel Planning                                  1
                     SaaS product intelligence platform                      1
Agra                 eCommerce                                               2
Ahmedabad            Consumer Internet                                      12
                     Technology                                              6
                     eCommerce                                               6
                     Cloud Based Collaboration platform                      1
                     Custom Merchandize platform                             1
                     Enterprise Communication Platform     

**Grouping w.r.t. multiple columns**

In [90]:
df.groupby(['CityLocation', 'InvestmentType'])['AmountInUSD'].sum()

CityLocation           InvestmentType
0                      Crowd Funding     1.557680e+05
                       Private Equity    1.237675e+09
                       Seed Funding      3.403310e+07
Agra                   Seed Funding      0.000000e+00
Ahmedabad              Debt Funding      7.800000e+06
                       Private Equity    8.565000e+07
                       Seed Funding      4.736000e+06
Bangalore              Private Equity    8.289120e+09
                       Seed Funding      9.465411e+07
Bangalore / Palo Alto  Seed Funding      1.000000e+06
Bangalore / SFO        Private Equity    1.350000e+07
                       Seed Funding      1.800000e+06
Bangalore / San Mateo  Private Equity    8.000000e+06
Bangalore / USA        Private Equity    5.000000e+06
Bangalore/ Bangkok     Private Equity    9.900000e+06
Belgaum                Seed Funding      5.000000e+05
Bhopal                 Private Equity    1.800000e+06
                       Seed Funding      1.0

## Resetting Index

In [91]:
type(df.groupby(['CityLocation', 'InvestmentType'])['AmountInUSD'].sum())

pandas.core.series.Series

In [92]:
df.groupby(['CityLocation', 'InvestmentType'])['AmountInUSD'].sum().reset_index()

CityLocation  InvestmentType   AmountInUSD
0                       0   Crowd Funding  1.557680e+05
1                       0  Private Equity  1.237675e+09
2                       0    Seed Funding  3.403310e+07
3                    Agra    Seed Funding  0.000000e+00
4               Ahmedabad    Debt Funding  7.800000e+06
5               Ahmedabad  Private Equity  8.565000e+07
6               Ahmedabad    Seed Funding  4.736000e+06
7               Bangalore  Private Equity  8.289120e+09
8               Bangalore    Seed Funding  9.465411e+07
9   Bangalore / Palo Alto    Seed Funding  1.000000e+06
10        Bangalore / SFO  Private Equity  1.350000e+07
11        Bangalore / SFO    Seed Funding  1.800000e+06
12  Bangalore / San Mateo  Private Equity  8.000000e+06
13        Bangalore / USA  Private Equity  5.000000e+06
14     Bangalore/ Bangkok  Private Equity  9.900000e+06
15                Belgaum    Seed Funding  5.000000e+05
16                 Bhopal  Private Equity  1.800000e+06
17                 Bhopal    Seed Funding  1.000000e+05
18                 Boston  Private Equity  3.000000e+06
19             Chandigarh  Private Equity  2.500000e+07
20             Chandigarh    Seed Funding  1.100000e+06
21                Chennai  Private Equity  4.020710e+08
22                Chennai    Seed Funding  9.034000e+06
23     Chennai/ Singapore  Private Equity  0.000000e+00
24             Coimbatore    Seed Funding  1.650000e+06
25     Dallas / Hyderabad    Seed Funding  1.000000e+06
26                  Delhi  Private Equity  1.000000e+06
27                  Delhi    Seed Funding  0.000000e+00
28                    Goa  Private Equity  2.000000e+06
29                    Goa    Seed Funding  3.800000e+05
..                    ...             ...           ...
69                  Noida  Private Equity  1.590000e+08
70                  Noida    Seed Funding  1.133800e+07
71      Noida / Singapore    Seed Funding  3.000000e+05
72                 Panaji    Seed Funding  8.250000e+05
73                   Pune  Private Equity  2.731550e+08
74                   Pune    Seed Funding  8.998000e+06
75           Pune / Dubai    Seed Funding  1.000000e+06
76       Pune / Singapore  Private Equity  0.000000e+00
77              Pune / US  Private Equity  6.850000e+07
78           Pune/Seattle  Private Equity  1.500000e+07
79        SFO / Bangalore  Private Equity  2.700000e+06
80    Seattle / Bangalore    Seed Funding  0.000000e+00
81               Siliguri  Private Equity  0.000000e+00
82               Siliguri    Seed Funding  0.000000e+00
83              Singapore  Private Equity  3.000000e+06
84              Singapore    Seed Funding  8.500000e+05
85                  Surat    Seed Funding  0.000000e+00
86             Trivandrum  Private Equity  0.000000e+00
87             Trivandrum    Seed Funding  1.000000e+05
88                     US    Seed Funding  0.000000e+00
89               US/India  Private Equity  3.000000e+06
90                    USA  Private Equity  0.000000e+00
91              USA/India  Private Equity  1.660000e+07
92                Udaipur  Private Equity  0.000000e+00
93                  Udupi  Private Equity  1.200000e+07
94                  Udupi    Seed Funding  0.000000e+00
95               Vadodara  Private Equity  5.000000e+06
96               Vadodara    Seed Funding  1.040000e+06
97               Varanasi    Seed Funding  5.200000e+04
98              bangalore    Seed Funding  0.000000e+00

[99 rows x 3 columns]

In [93]:
type(df.groupby(['CityLocation', 'InvestmentType'])['AmountInUSD'].sum().reset_index())

pandas.core.frame.DataFrame

In [94]:
grouped=df.groupby(['CityLocation', 'InvestmentType'])
grouped.agg({'AmountInUSD':['min','max','mean']})
grouped.agg({'AmountInUSD':{'min':np.min,'mean':np.mean,'max':np.max}})

C:\ProgramData\Anaconda3\lib\site-packages\pandas\core\groupby.py:4291: FutureWarning: using a dict with renaming is deprecated and will be removed in a future version
  return super(DataFrameGroupBy, self).aggregate(arg, *args, **kwargs)


AmountInUSD                          
                                             min          mean         max
CityLocation          InvestmentType                                      
0                     Crowd Funding        30768  7.788400e+04      125000
                      Private Equity           0  1.422615e+07   110000000
                      Seed Funding             0  3.781456e+05     3250000
Agra                  Seed Funding             0  0.000000e+00           0
Ahmedabad             Debt Funding       7800000  7.800000e+06     7800000
                      Private Equity     1000000  7.786364e+06    32000000
                      Seed Funding             0  2.059130e+05     1000000
Bangalore             Private Equity           0  2.726684e+07  1400000000
                      Seed Funding             0  2.930468e+05     7500000
Bangalore / Palo Alto Seed Funding       1000000  1.000000e+06     1000000
Bangalore / SFO       Private Equity    13500000  1.350000e+07    13500000
                      Seed Funding       1800000  1.800000e+06     1800000
Bangalore / San Mateo Private Equity     8000000  8.000000e+06     8000000
Bangalore / USA       Private Equity     5000000  5.000000e+06     5000000
Bangalore/ Bangkok    Private Equity     1900000  4.950000e+06     8000000
Belgaum               Seed Funding        500000  5.000000e+05      500000
Bhopal                Private Equity     1800000  1.800000e+06     1800000
                      Seed Funding        100000  1.000000e+05      100000
Boston                Private Equity     3000000  3.000000e+06     3000000
Chandigarh            Private Equity           0  5.000000e+06    10000000
                      Seed Funding             0  2.200000e+05     1000000
Chennai               Private Equity           0  1.148774e+07    60000000
                      Seed Funding             0  2.914194e+05     1500000
Chennai/ Singapore    Private Equity           0  0.000000e+00           0
Coimbatore            Seed Funding             0  5.500000e+05     1200000
Dallas / Hyderabad    Seed Funding       1000000  1.000000e+06     1000000
Delhi                 Private Equity           0  5.000000e+05     1000000
                      Seed Funding             0  0.000000e+00           0
Goa                   Private Equity           0  6.666667e+05     1000000
                      Seed Funding             0  9.500000e+04      200000
...                                          ...           ...         ...
Noida                 Private Equity           0  4.676471e+06    40000000
                      Seed Funding             0  2.576818e+05     2000000
Noida / Singapore     Seed Funding        300000  3.000000e+05      300000
Panaji                Seed Funding        825000  8.250000e+05      825000
Pune                  Private Equity           0  8.033971e+06    56000000
                      Seed Funding             0  1.799600e+05     1000000
Pune / Dubai          Seed Funding       1000000  1.000000e+06     1000000
Pune / Singapore      Private Equity           0  0.000000e+00           0
Pune / US             Private Equity     6000000  1.712500e+07    35000000
Pune/Seattle          Private Equity    15000000  1.500000e+07    15000000
SFO / Bangalore       Private Equity     2700000  2.700000e+06     2700000
Seattle / Bangalore   Seed Funding             0  0.000000e+00           0
Siliguri              Private Equity           0  0.000000e+00           0
                      Seed Funding             0  0.000000e+00           0
Singapore             Private Equity     3000000  3.000000e+06     3000000
                      Seed Funding             0  2.125000e+05      500000
Surat                 Seed Funding             0  0.000000e+00           0
Trivandrum            Private Equity           0  0.000000e+00           0
                      Seed Funding        100000  1.000000e+05      100000
US                    Seed Funding             0  0.000000e+00

In [95]:
df.dtypes

SNo                          int64
Date                datetime64[ns]
StartupName                 object
IndustryVertical            object
SubVertical                 object
CityLocation                object
InvestorsName               object
InvestmentType              object
AmountInUSD                  int32
Remarks                     object
month                      float64
dtype: object

In [96]:
new = df.groupby(['CityLocation', 'InvestmentType'])['AmountInUSD'].sum().reset_index()
new['CityLocation'] = new['CityLocation'].str.replace('0', 'Other')
new

CityLocation  InvestmentType   AmountInUSD
0                   Other   Crowd Funding  1.557680e+05
1                   Other  Private Equity  1.237675e+09
2                   Other    Seed Funding  3.403310e+07
3                    Agra    Seed Funding  0.000000e+00
4               Ahmedabad    Debt Funding  7.800000e+06
5               Ahmedabad  Private Equity  8.565000e+07
6               Ahmedabad    Seed Funding  4.736000e+06
7               Bangalore  Private Equity  8.289120e+09
8               Bangalore    Seed Funding  9.465411e+07
9   Bangalore / Palo Alto    Seed Funding  1.000000e+06
10        Bangalore / SFO  Private Equity  1.350000e+07
11        Bangalore / SFO    Seed Funding  1.800000e+06
12  Bangalore / San Mateo  Private Equity  8.000000e+06
13        Bangalore / USA  Private Equity  5.000000e+06
14     Bangalore/ Bangkok  Private Equity  9.900000e+06
15                Belgaum    Seed Funding  5.000000e+05
16                 Bhopal  Private Equity  1.800000e+06
17                 Bhopal    Seed Funding  1.000000e+05
18                 Boston  Private Equity  3.000000e+06
19             Chandigarh  Private Equity  2.500000e+07
20             Chandigarh    Seed Funding  1.100000e+06
21                Chennai  Private Equity  4.020710e+08
22                Chennai    Seed Funding  9.034000e+06
23     Chennai/ Singapore  Private Equity  0.000000e+00
24             Coimbatore    Seed Funding  1.650000e+06
25     Dallas / Hyderabad    Seed Funding  1.000000e+06
26                  Delhi  Private Equity  1.000000e+06
27                  Delhi    Seed Funding  0.000000e+00
28                    Goa  Private Equity  2.000000e+06
29                    Goa    Seed Funding  3.800000e+05
..                    ...             ...           ...
69                  Noida  Private Equity  1.590000e+08
70                  Noida    Seed Funding  1.133800e+07
71      Noida / Singapore    Seed Funding  3.000000e+05
72                 Panaji    Seed Funding  8.250000e+05
73                   Pune  Private Equity  2.731550e+08
74                   Pune    Seed Funding  8.998000e+06
75           Pune / Dubai    Seed Funding  1.000000e+06
76       Pune / Singapore  Private Equity  0.000000e+00
77              Pune / US  Private Equity  6.850000e+07
78           Pune/Seattle  Private Equity  1.500000e+07
79        SFO / Bangalore  Private Equity  2.700000e+06
80    Seattle / Bangalore    Seed Funding  0.000000e+00
81               Siliguri  Private Equity  0.000000e+00
82               Siliguri    Seed Funding  0.000000e+00
83              Singapore  Private Equity  3.000000e+06
84              Singapore    Seed Funding  8.500000e+05
85                  Surat    Seed Funding  0.000000e+00
86             Trivandrum  Private Equity  0.000000e+00
87             Trivandrum    Seed Funding  1.000000e+05
88                     US    Seed Funding  0.000000e+00
89               US/India  Private Equity  3.000000e+06
90                    USA  Private Equity  0.000000e+00
91              USA/India  Private Equity  1.660000e+07
92                Udaipur  Private Equity  0.000000e+00
93                  Udupi  Private Equity  1.200000e+07
94                  Udupi    Seed Funding  0.000000e+00
95               Vadodara  Private Equity  5.000000e+06
96               Vadodara    Seed Funding  1.040000e+06
97               Varanasi    Seed Funding  5.200000e+04
98              bangalore    Seed Funding  0.000000e+00

[99 rows x 3 columns]

## Pivoting & Melting

**Pivoting**

In [97]:
new.fillna('0',inplace=True)
new

CityLocation  InvestmentType   AmountInUSD
0                   Other   Crowd Funding  1.557680e+05
1                   Other  Private Equity  1.237675e+09
2                   Other    Seed Funding  3.403310e+07
3                    Agra    Seed Funding  0.000000e+00
4               Ahmedabad    Debt Funding  7.800000e+06
5               Ahmedabad  Private Equity  8.565000e+07
6               Ahmedabad    Seed Funding  4.736000e+06
7               Bangalore  Private Equity  8.289120e+09
8               Bangalore    Seed Funding  9.465411e+07
9   Bangalore / Palo Alto    Seed Funding  1.000000e+06
10        Bangalore / SFO  Private Equity  1.350000e+07
11        Bangalore / SFO    Seed Funding  1.800000e+06
12  Bangalore / San Mateo  Private Equity  8.000000e+06
13        Bangalore / USA  Private Equity  5.000000e+06
14     Bangalore/ Bangkok  Private Equity  9.900000e+06
15                Belgaum    Seed Funding  5.000000e+05
16                 Bhopal  Private Equity  1.800000e+06
17                 Bhopal    Seed Funding  1.000000e+05
18                 Boston  Private Equity  3.000000e+06
19             Chandigarh  Private Equity  2.500000e+07
20             Chandigarh    Seed Funding  1.100000e+06
21                Chennai  Private Equity  4.020710e+08
22                Chennai    Seed Funding  9.034000e+06
23     Chennai/ Singapore  Private Equity  0.000000e+00
24             Coimbatore    Seed Funding  1.650000e+06
25     Dallas / Hyderabad    Seed Funding  1.000000e+06
26                  Delhi  Private Equity  1.000000e+06
27                  Delhi    Seed Funding  0.000000e+00
28                    Goa  Private Equity  2.000000e+06
29                    Goa    Seed Funding  3.800000e+05
..                    ...             ...           ...
69                  Noida  Private Equity  1.590000e+08
70                  Noida    Seed Funding  1.133800e+07
71      Noida / Singapore    Seed Funding  3.000000e+05
72                 Panaji    Seed Funding  8.250000e+05
73                   Pune  Private Equity  2.731550e+08
74                   Pune    Seed Funding  8.998000e+06
75           Pune / Dubai    Seed Funding  1.000000e+06
76       Pune / Singapore  Private Equity  0.000000e+00
77              Pune / US  Private Equity  6.850000e+07
78           Pune/Seattle  Private Equity  1.500000e+07
79        SFO / Bangalore  Private Equity  2.700000e+06
80    Seattle / Bangalore    Seed Funding  0.000000e+00
81               Siliguri  Private Equity  0.000000e+00
82               Siliguri    Seed Funding  0.000000e+00
83              Singapore  Private Equity  3.000000e+06
84              Singapore    Seed Funding  8.500000e+05
85                  Surat    Seed Funding  0.000000e+00
86             Trivandrum  Private Equity  0.000000e+00
87             Trivandrum    Seed Funding  1.000000e+05
88                     US    Seed Funding  0.000000e+00
89               US/India  Private Equity  3.000000e+06
90                    USA  Private Equity  0.000000e+00
91              USA/India  Private Equity  1.660000e+07
92                Udaipur  Private Equity  0.000000e+00
93                  Udupi  Private Equity  1.200000e+07
94                  Udupi    Seed Funding  0.000000e+00
95               Vadodara  Private Equity  5.000000e+06
96               Vadodara    Seed Funding  1.040000e+06
97               Varanasi    Seed Funding  5.200000e+04
98              bangalore    Seed Funding  0.000000e+00

[99 rows x 3 columns]

In [107]:
pivoted=pd.pivot_table(new,index='InvestmentType', columns='CityLocation', values='AmountInUSD',aggfunc='sum',fill_value=0)

In [108]:
pivoted

CityLocation    Agra  Ahmedabad   Bangalore  Bangalore / Palo Alto  \
InvestmentType                                                       
Crowd Funding      0          0           0                      0   
Debt Funding       0    7800000           0                      0   
Other              0          0           0                      0   
Private Equity     0   85650000  8289120000                      0   
Seed Funding       0    4736000    94654108                1000000   

CityLocation    Bangalore / SFO  Bangalore / San Mateo  Bangalore / USA  \
InvestmentType                                                            
Crowd Funding                 0                      0                0   
Debt Funding                  0                      0                0   
Other                         0                      0                0   
Private Equity         13500000                8000000          5000000   
Seed Funding            1800000                      0                0   

CityLocation    Bangalore/ Bangkok  Belgaum   Bhopal    ...      Trivandrum  \
InvestmentType                                          ...                   
Crowd Funding                    0        0        0    ...               0   
Debt Funding                     0        0        0    ...               0   
Other                            0        0        0    ...               0   
Private Equity             9900000        0  1800000    ...               0   
Seed Funding                     0   500000   100000    ...          100000   

CityLocation    US  US/India  USA  USA/India  Udaipur     Udupi  Vadodara  \
InvestmentType                                                              
Crowd Funding    0         0    0          0        0         0         0   
Debt Funding     0         0    0          0        0         0         0   
Other            0         0    0          0        0         0         0   
Private Equity   0   3000000    0   16600000        0  12000000   5000000   
Seed Funding     0         0    0          0        0         0   1040000   

CityLocation    Varanasi  bangalore  
InvestmentType                       
Crowd Funding          0          0  
Debt Funding           0          0  
Other                  0          0  
Private Equity         0          0  
Seed Funding       52000          0  

[5 rows x 72 columns]

In [111]:
a=pivoted.reset_index()

**Melting**

In [112]:
pd.melt(a, id_vars=['InvestmentType'], value_vars=['Agra','Ahmedabad'], value_name='Amount')

InvestmentType CityLocation    Amount
0   Crowd Funding         Agra         0
1    Debt Funding         Agra         0
2           Other         Agra         0
3  Private Equity         Agra         0
4    Seed Funding         Agra         0
5   Crowd Funding    Ahmedabad         0
6    Debt Funding    Ahmedabad   7800000
7           Other    Ahmedabad         0
8  Private Equity    Ahmedabad  85650000
9    Seed Funding    Ahmedabad   4736000

## Column / Row Concatenation

In [113]:
df.columns

Index(['SNo', 'Date', 'StartupName', 'IndustryVertical', 'SubVertical',
       'CityLocation', 'InvestorsName', 'InvestmentType', 'AmountInUSD',
       'Remarks', 'month'],
      dtype='object')

In [114]:
part1 = df[['SNo', 'Date', 'StartupName']]
part2 = df[['InvestorsName', 'InvestmentType']]
print(part1.shape)
part1.head()

(2372, 3)


SNo       Date   StartupName
0    0 2017-08-01      TouchKin
1    1 2017-08-02       Ethinos
2    2 2017-08-02  Leverage Edu
3    3 2017-08-02          Zepo
4    4 2017-08-02  Click2Clinic

In [115]:
print(part2.shape)
part2.head()

(2372, 2)


InvestorsName  InvestmentType
0                                        Kae Capital  Private Equity
1                         Triton Investment Advisors  Private Equity
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding
3  Kunal Shah, LetsVenture, Anupam Mittal, Hetal ...    Seed Funding
4                     Narottam Thudi, Shireesh Palle    Seed Funding

In [116]:
pd.concat([part1, part2], axis=1)

SNo       Date       StartupName  \
0        0 2017-08-01          TouchKin   
1        1 2017-08-02           Ethinos   
2        2 2017-08-02      Leverage Edu   
3        3 2017-08-02              Zepo   
4        4 2017-08-02      Click2Clinic   
5        5 2017-07-01     Billion Loans   
6        6 2017-07-03  Ecolibriumenergy   
7        7 2017-07-04             Droom   
8        8 2017-07-05         Jumbotail   
9        9 2017-07-05            Moglix   
10      10 2017-07-05        Timesaverz   
11      11 2017-07-06            Minjar   
12      12 2017-07-06       MyCity4kids   
13      13 2017-07-07          Clip App   
14      14 2017-07-07       Upwardly.in   
15      15 2017-07-10        Autorox.co   
16      16 2017-07-11            Fabogo   
17      17 2017-07-11        Flickstree   
18      18 2017-07-11       Design Cafe   
19      19 2017-07-12          Innoviti   
20      20 2017-07-12          VDeliver   
21      21 2017-07-12          Bottr.me   
22      22 2017-07-12          Arcatron   
23      23 2017-07-14          QwikSpec   
24      24 2017-07-14           Chumbak   
25      25 2017-07-17           Increff   
26      26 2017-07-17            Vayana   
27      27 2017-07-18         MObiquest   
28      28 2017-07-18             Ambee   
29      29 2017-07-18   Ideal Insurance   
...    ...        ...               ...   
2342  2342 2015-01-19           Refiral   
2343  2343 2015-01-19      Ping Digital   
2344  2344 2015-01-19   HealthyWorld.in   
2345  2345 2015-01-19      Hostel Dunia   
2346  2346 2015-01-19   Play your sport   
2347  2347 2015-01-19           Yo Grad   
2348  2348 2015-01-19            Enabli   
2349  2349 2015-01-19               CBS   
2350  2350 2015-01-20         Hopscotch   
2351  2351 2015-01-20  Amagi Media Labs   
2352  2352 2015-01-20   HealthyWorld.in   
2353  2353 2015-01-21       Simplilearn   
2354  2354 2015-01-21      MyDreamStore   
2355  2355 2015-05-21              Knit   
2356  2356 2015-05-21           Villgro   
2357  2357 2015-01-22      Corporate360   
2358  2358 2015-01-22         Freshmonk   
2359  2359 2015-01-22   Englishleap.com   
2360  2360 2015-01-22            Antuit   
2361  2361 2015-01-22     Caratlane.com   
2362  2362 2015-01-24        Aasaanjobs   
2363  2363 2015-01-24          Impartus   
2364  2364 2015-01-25        Thrill App   
2365  2365 2015-01-27      CarDekho.com   
2366  2366 2015-01-28     Grabhouse.com   
2367  2367 2015-01-29        Printvenue   
2368  2368 2015-01-29          Graphene   
2369  2369 2015-01-30    Mad Street Den   
2370  2370 2015-01-30         Simplotel   
2371  2371 2015-01-31  couponmachine.in   

                                          InvestorsName  InvestmentType  
0                                           Kae Capital  Private Equity  
1                            Triton Investment Advisors  Private Equity  
2     Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding  
3     Kunal Shah, LetsVenture, Anupam Mittal, Hetal ...    Seed Funding  
4                        Narottam Thudi, Shireesh Palle    Seed Funding  
5              Reliance Corporate Advisory Services Ltd    Seed Funding  
6                                  Infuse Ventures, JLL  Private Equity  
7       Asset Management (Asia) Ltd, Digital Garage Inc  Private Equity  
8         Kalaari Capital, Nexus India Capital Advisors  Private Equity  
9     International Finance Corporation, Rocketship,...  Private Equity  
10                                                 BCCL  Private Equity  
11    Blume Ventures, Contrarian Capital India Partn...    Seed Funding  
12                                                    0    Seed Funding  
13                      India Quotient, Shunwei Capital    Seed Funding  
14    Sreeram Iyer, Suvo Sarkar, Anita Gupta, Likemi...    Seed Funding  
15                                         Mitsui & Co.  Private Equity  
16                             Dunamis Ventures Pte Ltd  Private Equity  
17    Ve

**Rows Concatenation**

In [117]:
rows1 = df[:3]
rows2 = df[5:7]
print(rows1.shape)
rows1

(3, 11)


SNo       Date   StartupName   IndustryVertical  \
0    0 2017-08-01      TouchKin         Technology   
1    1 2017-08-02       Ethinos         Technology   
2    2 2017-08-02  Leverage Edu  Consumer Internet   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   

   AmountInUSD Remarks  month  
0      1300000       0    8.0  
1            0       0    8.0  
2            0       0    8.0

In [118]:
print(rows2.shape)
rows2

(2, 11)


SNo       Date       StartupName   IndustryVertical  \
5    5 2017-07-01     Billion Loans  Consumer Internet   
6    6 2017-07-03  Ecolibriumenergy         Technology   

                            SubVertical CityLocation  \
5         Peer to Peer Lending platform    Bangalore   
6  Energy management solutions provider    Ahmedabad   

                              InvestorsName  InvestmentType  AmountInUSD  \
5  Reliance Corporate Advisory Services Ltd    Seed Funding      1000000   
6                      Infuse Ventures, JLL  Private Equity      2600000   

  Remarks  month  
5       0    7.0  
6       0    7.0

In [119]:
pd.concat([rows1, rows2], axis=0)

SNo       Date       StartupName   IndustryVertical  \
0    0 2017-08-01          TouchKin         Technology   
1    1 2017-08-02           Ethinos         Technology   
2    2 2017-08-02      Leverage Edu  Consumer Internet   
5    5 2017-07-01     Billion Loans  Consumer Internet   
6    6 2017-07-03  Ecolibriumenergy         Technology   

                                     SubVertical CityLocation  \
0                       Predictive Care Platform    Bangalore   
1                       Digital Marketing Agency       Mumbai   
2  Online platform for Higher Education Services    New Delhi   
5                  Peer to Peer Lending platform    Bangalore   
6           Energy management solutions provider    Ahmedabad   

                                       InvestorsName  InvestmentType  \
0                                        Kae Capital  Private Equity   
1                         Triton Investment Advisors  Private Equity   
2  Kashyap Deorah, Anand Sankeshwar, Deepak Jain,...    Seed Funding   
5           Reliance Corporate Advisory Services Ltd    Seed Funding   
6                               Infuse Ventures, JLL  Private Equity   

   AmountInUSD Remarks  month  
0      1300000       0    8.0  
1            0       0    8.0  
2            0       0    8.0  
5      1000000       0    7.0  
6      2600000       0    7.0

## Looping through DataFrame

#### Using Simple for loop

In [120]:
import time
start = time.time()
for x in df['StartupName']:
    print(x)
end = time.time()
print(end - start)

TouchKin
Ethinos
Leverage Edu
Zepo
Click2Clinic
Billion Loans
Ecolibriumenergy
Droom
Jumbotail
Moglix
Timesaverz
Minjar
MyCity4kids
Clip App
Upwardly.in
Autorox.co
Fabogo
Flickstree
Design Cafe
Innoviti
VDeliver
Bottr.me
Arcatron
QwikSpec
Chumbak
Increff
Vayana
MObiquest
Ambee
Ideal Insurance
Hypernova Interactive
Rentomojo
AirCTO
Playablo
Trupay
Brick2Wall
FableStreet
Monsoon Fintech
MonkeyBox
Noticeboard
Byju’s
Creator’s Gurukul
Fab Hotels
ThinkerBell
1mg
Jhakaas
BigStylist
Gympik.com
Tripeur
RentOnGo
Goomo
MaxMyWealth
Spinny
Healthbuds
Ftcash
BHIVE Workspace
wayForward
GyanDhan
Hungry Foal
ZipLoan
GrowFitter
Stratfit
Multiplier Solutions
ABI Health
Stockal
Guidoo
eSahai.in
Loanmeet
LendingKart
Boxx.ai
PeeSafe.in
Vista Rooms
CoverFox
HyperTrack
Goodera
Digit
Ola
Vanitycask
MrNeeds
MoneyTap
Goodbox
Fincash
PortDesk
EdgeFx
EazyDiner
HealthIntel
Unbxd
DarwinBox
ForeverShop
Insider.in
Fynd
Grow Fit
Fabulyst
mSwipe
OneAssist
Julia Computing
Kissht
Explore Life traveling
Loantap
Voxweb
HUG

Ixigo
SkillAngels
Vidgyor
Banihal
Goodservice
Treebo
EduKart
Healthians.com
LabInApp
Syona Cosmetics
Sigrid Education
Indix
The Porter
Retention.ai
Fusion Microfinance
Zinka
CreditMantri
GetMyPeon
UrbanClap
StayGlad
AdWyze
Zimmber
Buttercups
RailYatri
aagaar.com
Flatchat
MyCFO
Foodpanda
Termsheet
Applicate
World Art Community
SpoonJoy
Seed Schools
Ignis Careers
Square Yards
Fintellix
Customer360
Delhivery
Swiggy
Uniphore
Box8
Toppr
Vedantu
BuyHatke
KleverKid
Uniken
GrandOpinion
HealthifyMe
ZapStitch
Wassup
Mobiefit
Plancess
WorkHorse
Innovaccer
MyCuteOffice
IndianRoots.com
RedPolka
Venturesity
TheBetterIndia.com
Awaaz De
Chaayos
Akosha
Cooey
MeetUniv
Inspirock
Crown-it
Trucksfirst
Lookup
SunTerrace
Nudgespot
CarDekho
EveningFlavors
Razorpay
ZenParent
Newgen Payments
UrbanPro
Goodbox
Renderlogy
Postman
Casa2inns
FleetRover
TheKarrier
Zoomcar
Truweight
Ather Energy
Swiggy
Bluegape
KeepTrax
InstaLively
Pricejugaad
Quikr
PressPlay
LogiNext
FirstCry.com
MobiKwik
Olacabs
Gadgetwood
Bonhomia


#### Using iterrows func

In [121]:

start = time.time()

for index, row in df.iterrows():
    
    print(index,row['StartupName'], row['CityLocation'])
end = time.time()
print(end - start)

0 TouchKin Bangalore
1 Ethinos Mumbai
2 Leverage Edu New Delhi
3 Zepo Mumbai
4 Click2Clinic Hyderabad
5 Billion Loans Bangalore
6 Ecolibriumenergy Ahmedabad
7 Droom Gurgaon
8 Jumbotail Bangalore
9 Moglix Noida
10 Timesaverz Mumbai
11 Minjar Bangalore
12 MyCity4kids Gurgaon
13 Clip App Bangalore
14 Upwardly.in Bangalore
15 Autorox.co Hyderabad
16 Fabogo Pune
17 Flickstree Mumbai
18 Design Cafe Bangalore
19 Innoviti Bangalore
20 VDeliver Hyderabad
21 Bottr.me Bangalore
22 Arcatron Pune
23 QwikSpec Bangalore
24 Chumbak Bangalore
25 Increff Bangalore
26 Vayana Pune
27 MObiquest Noida
28 Ambee Hyderabad
29 Ideal Insurance Kolkata
30 Hypernova Interactive Bangalore
31 Rentomojo Bangalore
32 AirCTO Bangalore
33 Playablo Bangalore
34 Trupay Gurgaon
35 Brick2Wall Gurgaon
36 FableStreet New Delhi
37 Monsoon Fintech New Delhi
38 MonkeyBox Bangalore
39 Noticeboard Bangalore
40 Byju’s Bangalore
41 Creator’s Gurukul New Delhi
42 Fab Hotels New Delhi
43 ThinkerBell Bangalore
44 1mg Gurgaon
45 Jhakaas

491 Zoomcar Bangalore
492 BYJU’s Bangalore
493 Supr Daily Mumbai
494 The BlueBook Bangalore
495 MamaEarth New Delhi
496 Vegetall Chennai
497 Innoplexus Pune
498 Mindler New Delhi
499 MCaffeine Bangalore
500 enKast Bangalore
501 ShopKirana Indore
502 MyAdvo New Delhi
503 Asocon Noida
504 LexComply New Delhi
505 The Postbox Chennai
506 Pandorum Technologies Bangalore
507 Dekkho Mumbai
508 Shadowfox Gurgaon
509 ShopX Bangalore
510 BookEventz Mumbai
511 Healthifyme Bangalore
512 LetsMD New Delhi
513 Justbooks Bangalore
514 InstantPay New Delhi
515 PickMyLaundry New Delhi
516 FreshDesk Chennai
517 LaundryAnna Bangalore
518 LetsMD New Delhi
519 Sattviko New Delhi
520 Zarget Chennai
521 DoSelect Bangalore
522 EasyPolicy Noida
523 Browntape Goa
524 Connect India Bangalore
525 Corseco New Delhi
526 vImmune Bangalore
527 The Gourmet Jar Noida
528 Medzin New Delhi
529 Bizongo Mumbai
530 Rivigo Gurgaon
531 JustRide Mumbai
532 IIM Jobs New Delhi
533 Square Yards Gurgaon
534 GolfLAN New Delhi
535 Yo

891 Chikoop Indore
892 BRIDGEi2I Bangalore
893 Droom New Delhi
894 Visit New Delhi
895 Atomberg Mumbai
896 Coutloot Mumbai
897 Limo Mumbai
898 iCliniq Bangalore
899 GoGo Truck Chennai
900 Vyome Biosciences New Delhi
901 SillyMonks Hyderabad
902 CueMath Bangalore
903 KrazyBee Bangalore
904 Intgrea Bangalore
905 HelpShift Pune
906 CreditVidya Mumbai
907 Grey Campus Hyderabad
908 ListUp Mumbai
909 6Degree Mumbai
910 Shopwati Gurgaon
911 Cookifi Bangalore
912 StoreKing Bangalore
913 Wired Hub Jaipur
914 MySeniorDoctor Gurgaon
915 Teabox Siliguri
916 ExtraaEdge Pune
917 EPayLater Mumbai
918 Kyazoonga Mumbai
919 Paytunes New Delhi
920 Kickstart Jobs Gurgaon
921 Freshee Mumbai
922 MadRat Games Bangalore
923 Jazzmyride New Delhi
924 AutoSense New Delhi
925 Redsun Telematics Chennai
926 KhanaGadi Jaipur
927 Bombay Shirt Company Mumbai
928 Matrubharti 0
929 EdTechReview New Delhi
930 Bonhomia New Delhi
931 Gobolt New Delhi
932 Jivox Bangalore
933 Sminq Pune
934 CureJoy Bangalore
935 Oneway.cab V

1331 PinClick Bangalore
1332 Smartcooky New Delhi
1333 Helpi Mumbai
1334 DogSpot Gurgaon
1335 Career360 New Delhi
1336 B9 Beverages New Delhi
1337 WIMWI Foods Ahmedabad
1338 MyChild App Bhopal
1339 Instaproc Noida
1340 360Ride Bangalore
1341 AlefMobitech Mumbai
1342 Tracxn Bangalore
1343 Qdesq Gurgaon
1344 Sensara Bangalore
1345 Zebpay Ahmedabad
1346 PlanMyMedicalTrip Pune
1347 MyCity4Kids Gurgaon
1348 Justdakhila.com New Delhi
1349 PitStop Bangalore
1350 Bikxie Gurgaon
1351 Infurnia Bangalore
1352 Myly Jaipur
1353 MakeMyTrip Gurgaon
1354 Melorra Bangalore
1355 ConfirmTKT Bangalore
1356 Gyaanzone Mumbai
1357 FlatFurnish Gurgaon
1358 Lumiere Bangalore
1359 Koovs Gurgaon
1360 Wishup New Delhi
1361 FreshMenu Bangalore
1362 SavvyMob Bangalore
1363 Fashionablyin Mumbai
1364 Shopclues Gurgaon
1365 Hike Messenger New Delhi
1366 iBus Networks Bangalore
1367 Happy2Refer Mumbai
1368 CarTrade Mumbai
1369 Rentomo Bangalore
1370 Roder New Delhi
1371 Bucker Hyderabad
1372 Care24 Mumbai
1373 Obino Mu

1686 Parcelled Bangalore
1687 AlmaMapper Ahmedabad
1688 Flexing It New Delhi
1689 Netmeds.com Chennai
1690 Razorpay Jaipur
1691 Jombay Pune
1692 Happitoo Mumbai
1693 Care24 Mumbai
1694 Roadrunnr Bangalore
1695 Simpli5d Gurgaon
1696 ORIGA Leasing Mumbai
1697 Smartivity.in New Delhi
1698 UrDoorStep Bangalore
1699 Tavaga Mumbai
1700 Moglix Noida
1701 ZopHop Mumbai
1702 Niki.ai Bangalore
1703 Instavans Bangalore
1704 LiftO Mumbai
1705 TinyOwl Mumbai
1706 Toppr.com Mumbai
1707 Kleeto Gurgaon
1708 BeaconsTalk Mumbai
1709 Jugnoo Chandigarh
1710 TOFlo Mumbai
1711 FXMartIndia Chandigarh
1712 Stylecracker Mumbai
1713 Luxuryhues Gurgaon
1714 HolaChef Mumbai
1715 Zivame Bangalore
1716 Capillary Tech New Delhi
1717 Jobspire Bangalore
1718 MeraDoctor Mumbai
1719 Vistaar Finance Bangalore
1720 HiJinny Mumbai
1721 Clapsnslaps Mumbai
1722 Blubox Mumbai
1723 RoomCentral Bangalore
1724 YouthKiAwaaz New Delhi
1725 Prozo New Delhi
1726 Zomato Gurgaon
1727 Sahayog Dairy Bhopal
1728 MockBank Bangalore
1729 R

2057 Goodservice New Delhi
2058 Treebo Bangalore
2059 EduKart New Delhi
2060 Healthians.com Hyderabad
2061 LabInApp Bangalore
2062 Syona Cosmetics Chennai
2063 Sigrid Education Noida
2064 Indix 0
2065 The Porter Mumbai
2066 Retention.ai Bangalore
2067 Fusion Microfinance New Delhi
2068 Zinka Bangalore
2069 CreditMantri Chennai
2070 GetMyPeon Mumbai
2071 UrbanClap New Delhi
2072 StayGlad Bangalore
2073 AdWyze Bangalore
2074 Zimmber Mumbai
2075 Buttercups Bangalore
2076 RailYatri Bangalore
2077 aagaar.com Gurgaon
2078 Flatchat Bangalore
2079 MyCFO Mumbai
2080 Foodpanda Gurgaon
2081 Termsheet Chennai
2082 Applicate Bangalore
2083 World Art Community Gurgaon
2084 SpoonJoy Bangalore
2085 Seed Schools Hyderabad
2086 Ignis Careers Hyderabad
2087 Square Yards Gurgaon
2088 Fintellix Bangalore
2089 Customer360 Mumbai
2090 Delhivery Gurgaon
2091 Swiggy Bangalore
2092 Uniphore Chennai
2093 Box8 Mumbai
2094 Toppr Mumbai
2095 Vedantu Bangalore
2096 BuyHatke Bangalore
2097 KleverKid New Delhi
2098 Un